text/0104-string-parameterization-output-spec.md

```diff
@@ -0,0 +1,293 @@
+
+```
+GET http://someservice.com/users/*/info
+```
+
+**Note:** The parameterizer must preserve the original URL structure as much as possible. e.g., the URL `https://service.io/data/sentry/frontend/info` where both `sentry` and `frontend` are dynamic route segments, the parameterizer must return `https://service.io/data/*/*/info` rather than `https://service.io/data/*/info`.
+
+### Subdomains
+
+Subdomain can contain parameters, too. For example, in the URL `https://company.someservice.io/` the string `company` might correspond to a client name, and is a dynamic string.
+
+If a parameterizer determines that a section of a domain string is dynamic, **it must replace the segment with a single `*` character.**
+
+Given:
+
+```
+GET http://host1.someservice.com/users/
+```
+
+If the parameterizer determines that `host1` is a parameter, it must return:
+
+```
+GET http://*.someservice.com/users/
+```
+
+**Note:** The parameterizer must not alter domains or hostnames in any other way.
+
+### Port Numbers
+
+**If the URL contains a port number, the parameterizer should not parameterize it.**
+
+Given:
+
+```
+GET http://someservice.com:8080/
+```
+
+Return:
+
+```
+GET http://someservice.com:8080/
+```
+
+### Auth Strings
+
+**If the URL contains authentication strings, the parameterizer must replace them with a `*` character.**
+
+Given:
+
+```
+GET http://admin:password@someservice.com/
+```
+
+Return:
+
+```
+GET http://*:*@someservice.com/
+```
+
+## Parameterizing Key-Value Store Key Names
+
+Key-value stores like Redis store arbitrary data, and are accessed by a “key”, which is a string. It is a common convention to namespace these keys, delimit them with a `:` or `.` character, and put parameters between the namespaces. e.g., the order IDs of a given user might be stored under the key `organizations:18:users:9982:orders`. `18` and `9982` are parameters. `organizations`, `users`, and `orders` are namespaces.
+
+## String Keys
+
+**For parameters in key names, the parameterizer must replace each parameter with a single `*` character.**
+
+Given:
+
+```
+SADD orgs:17:emails jane@myorg.com
+```
+
+Return:
+
+```
+SADD orgs:*:emails jane@myorg.com
+```
+
+## Structured JSON
+
+Databases like MongoDB structure operations as JSON payloads. In structured JSON languages, **the parameterizer must replace every leaf parameter with a single unquoted `?` character**.
+
+Given:
+
+```
+{
+  status: {
+    $in: ["active"]
+  }
+}
+```
+
+Return:
+
+```
+{
+  status: {
+    $in: [?]
+  }
+}
+```
+
+## Storing Parameterization Results
+
+**Parameterizers must keep the original value as it was sent, and add the parameterized description as a separate key in the payload.** For example, when parameterizing the `description` key of a payload:
+
+```json
+{
+  "description": "POST example.com/user/123456",
+}
+```
+
+The parameterizer must output:
+
+```json
+{
+  "description": "POST example.com/user/123456",
+	"description.scrubbed": "POST example.com/user/%s"
```

Comment:
There is an existing protocol for string substitutions, that [the UI partially understands](https://github.com/getsentry/sentry/blob/200624923c0066ae77710aba593f574ad59bb5c6/static/app/components/tagsTable.spec.tsx#L6-L44). It might make sense to reuse that. Can't find any documentation on it, but here's an [example](https://github.com/getsentry/relay/blob/ec72a5d6330581fab71330d3fa278bcdf4ffe488/relay-general/src/pii/snapshots/relay_general__pii__convert__tests__sdk_client_ip_stripped.snap#L5-L26) and [spec](https://github.com/getsentry/relay/blob/ec72a5d6330581fab71330d3fa278bcdf4ffe488/relay-general/src/types/meta.rs#L412-L430).

Reply:
I should have picked a better example, sorry. The point is that `_meta` has an optional `"val"` field we can write the original value to:

https://github.com/getsentry/relay/blob/a127f96ee4870f9d676e3d0f39ab91ab8c557fcd/relay-general/src/types/meta.rs#L427-L429

Reply:
Ah interesting, thanks!